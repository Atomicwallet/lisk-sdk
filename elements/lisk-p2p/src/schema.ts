/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Schema } from '@liskhq/lisk-codec';

export const nodeInfoSchema = {
	$id: '/nodeInfo',
	type: 'object',
	properties: {
		networkId: {
			dataType: 'string',
			fieldNumber: 1,
		},
		networkVersion: {
			dataType: 'string',
			fieldNumber: 2,
		},
		port: {
			dataType: 'uint32',
			fieldNumber: 3,
		},
		nonce: {
			dataType: 'string',
			fieldNumber: 4,
		},
		advertiseAddress: {
			dataType: 'boolean',
			fieldNumber: 5,
		},
		os: {
			dataType: 'string',
			fieldNumber: 6,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 7,
		},
	},
	required: ['networkId', 'networkVersion', 'port', 'nonce'],
};

export const peerInfoSchema = {
	$id: '/peerInfo',
	type: 'object',
	properties: {
		ipAddress: {
			dataType: 'string',
			fieldNumber: 1,
		},
		port: {
			dataType: 'uint32',
			fieldNumber: 2,
		},
		networkId: {
			dataType: 'string',
			fieldNumber: 3,
		},
		networkVersion: {
			dataType: 'string',
			fieldNumber: 4,
		},
		nonce: {
			dataType: 'string',
			fieldNumber: 5,
		},
		os: {
			dataType: 'string',
			fieldNumber: 6,
		},
		height: {
			dataType: 'uint32',
			fieldNumber: 7,
		},
	},
	required: ['ipAddress', 'port'],
};

export const defaultRPCSchemas = {
	peerInfo: peerInfoSchema,
	nodeInfo: nodeInfoSchema,
};

export const mergeCustomSchema = (
	baseSchema: Schema,
	customSchema: Schema,
): Schema => ({
	$id: `${baseSchema.$id}/custom`,
	type: 'object',
	properties: {
		...baseSchema.properties,
		...customSchema.properties,
	},
});
