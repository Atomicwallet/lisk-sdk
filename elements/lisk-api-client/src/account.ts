/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { codec } from '@liskhq/lisk-codec';
import { Channel, RegisteredSchemas } from './types';

export class Account {
	private readonly _channel: Channel;
	private readonly _schemas: RegisteredSchemas;

	public constructor(channel: Channel, schemas: RegisteredSchemas) {
		this._channel = channel;
		this._schemas = schemas;
	}

	public async get(address: Buffer): Promise<Record<string, unknown>> {
		const accountHex: string = await this._channel.invoke('app:getAccount', {
			address: address.toString('hex'),
		});

		return this.decode(Buffer.from(accountHex, 'hex'));
	}

	public encode(input: Record<string, unknown>): Buffer {
		return codec.encode(this._schemas.account, input);
	}

	public decode(input: Buffer): Record<string, unknown> {
		return codec.decode(this._schemas.account, input);
	}

	public toJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.toJSON(this._schemas.account, account);
	}

	public fromJSON(account: Record<string, unknown>): Record<string, unknown> {
		return codec.fromJSON(this._schemas.account, account);
	}
}
