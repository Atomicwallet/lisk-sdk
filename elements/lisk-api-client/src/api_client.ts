/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { EventCallback, Channel, RegisteredSchemas, NodeInfo } from './types';
import { Node } from './node';

export class APIClient {
	private readonly _channel: Channel;
	private _schemas!: RegisteredSchemas;
	private _nodeInfo!: NodeInfo;
	private _node!: Node;

	public constructor(channel: Channel) {
		this._channel = channel;
	}

	public async init(): Promise<void> {
		this._node = new Node(this._channel);
		this._schemas = await this._channel.invoke<RegisteredSchemas>('app:getSchema');
		this._nodeInfo = await this._node.getNodeInfo();
	}

	public async disconnect(): Promise<void> {
		return this._channel.disconnect();
	}

	public async invoke<T = Record<string, unknown>>(
		actionName: string,
		params?: Record<string, unknown>,
	): Promise<T> {
		return this._channel.invoke(actionName, params);
	}

	public subscribe<T = Record<string, unknown>>(eventName: string, cb: EventCallback<T>): void {
		this._channel.subscribe(eventName, cb);
	}

	public get node(): Node {
		// FIXME: remove unnecessary code
		// eslint-disable-next-line no-console
		console.log(this._schemas, this._nodeInfo);
		return this._node;
	}
}
