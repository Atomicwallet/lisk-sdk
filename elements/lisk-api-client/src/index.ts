/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export { createWSClient } from './ws_client';
export { createClient } from './client';
export { createAPIClient } from './create_api_client';
export { IPCChannel } from './ipc_channel';
