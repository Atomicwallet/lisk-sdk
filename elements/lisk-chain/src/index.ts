/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Chain, EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK } from './chain';
const events = { EVENT_DELETE_BLOCK, EVENT_NEW_BLOCK };

export { Account } from './account';
export * from './block_reward';
export { Chain, events };
export { baseBlockSchema } from './schema';
export { StateStore } from './state_store';
export { Slots } from './slots';
