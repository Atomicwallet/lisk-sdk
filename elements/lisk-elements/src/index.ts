/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
export * as apiClient from '../../lisk-api-client';
export * as cryptography from '../../lisk-cryptography';
export * as p2p from '../../lisk-p2p';
export * as passphrase from '../../lisk-passphrase';
export * as transactionPool from '../../lisk-transaction-pool';
export * as transactions from '../../lisk-transactions';
export * as utils from '../../lisk-utils';
export * as tree from '../../lisk-tree';
export * as validator from '../../lisk-validator';
export * as codec from '../../lisk-codec';
export * as db from '../../lisk-db';
export * as chain from '../../lisk-chain';
export * as bft from '../../lisk-bft';
export * as genesis from '../../lisk-genesis';
