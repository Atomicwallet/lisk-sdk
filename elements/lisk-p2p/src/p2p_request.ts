/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { RPCResponseAlreadySentError } from './errors';
import { P2PResponsePacketBufferData } from './types';

export interface RequestOptions {
	readonly procedure: string;
	readonly data: unknown;
	readonly id: string;
	readonly rate: number;
	productivity: {
		requestCounter: number;
		responseCounter: number;
		responseRate: number;
		lastResponded: number;
	};
}

export class P2PRequest {
	private readonly _procedure: string;
	private readonly _data: unknown;
	private readonly _respondCallback: (
		responseError?: Error,
		responseData?: P2PResponsePacketBufferData,
	) => void;
	private readonly _peerId: string;
	private _wasResponseSent: boolean;
	private readonly _rate: number;

	public constructor(
		options: RequestOptions,
		respondCallback: (responseError?: Error, responseData?: unknown) => void,
	) {
		this._procedure = options.procedure;
		this._data = options.data;
		this._peerId = options.id;
		this._rate = options.rate;
		// eslint-disable-next-line no-param-reassign
		options.productivity.requestCounter += 1;
		this._respondCallback = (
			responseError?: Error,
			responsePacket?: P2PResponsePacketBufferData,
		): void => {
			if (this._wasResponseSent) {
				throw new RPCResponseAlreadySentError(
					`A response has already been sent for the request procedure <<${options.procedure}>>`,
				);
			}
			this._wasResponseSent = true;
			// We assume peer performed useful work and update peer response rate
			if (!responseError && responsePacket) {
				// eslint-disable-next-line no-param-reassign
				options.productivity.lastResponded = Date.now();
				// eslint-disable-next-line no-param-reassign
				options.productivity.responseCounter += 1;
			}
			// eslint-disable-next-line no-param-reassign
			options.productivity.responseRate =
				options.productivity.responseCounter / options.productivity.requestCounter;

			let responsePacketBufferData: Buffer | undefined;
			if (responsePacket?.data && typeof responsePacket?.data === 'string') {
				responsePacketBufferData = Buffer.from(responsePacket.data, 'binary');
			}
			respondCallback(responseError, responsePacketBufferData);
		};
		this._wasResponseSent = false;
	}

	public get procedure(): string {
		return this._procedure;
	}

	public get data(): unknown {
		return this._data;
	}

	public get rate(): number {
		return this._rate;
	}

	public get peerId(): string {
		return this._peerId;
	}

	public get wasResponseSent(): boolean {
		return this._wasResponseSent;
	}

	public end(responseData: Buffer): void {
		const responsePacket: P2PResponsePacketBufferData = {
			data: responseData,
			peerId: this.peerId,
		};
		this._respondCallback(undefined, responsePacket);
	}

	public error(responseError: Error): void {
		this._respondCallback(responseError);
	}
}
