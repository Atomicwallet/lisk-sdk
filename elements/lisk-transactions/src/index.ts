/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */

import { DelegateTransaction } from './10_delegate_transaction';
import { MultisignatureTransaction } from './12_multisignature_transaction';
import { VoteTransaction } from './13_vote_transaction';
import { UnlockTransaction } from './14_unlock_transaction';
import { ProofOfMisbehaviorTransaction } from './15_proof_of_misbehavior_transaction';
import { TransferTransaction } from './8_transfer_transaction';
import { BaseTransaction, StateStore } from './base_transaction';
import * as constants from './constants';
import { convertToAssetError, convertToTransactionError, TransactionError } from './errors';
import { createResponse, Status, TransactionResponse } from './response';
import { transactionInterface } from './schema';
import {
	getSigningBytes,
	validateTransactionSchema,
	signTransaction,
	signMultiSignatureTransaction,
} from './transaction_helper';
import { Account, TransactionJSON } from './types';
import {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSenderIdAndPublicKey,
	validateSignature,
	verifyMinRemainingBalance,
	verifyMultiSignatureTransaction,
} from './utils';

const exposedUtils = {
	convertBeddowsToLSK,
	convertLSKToBeddows,
	prependMinusToPublicKeys,
	prependPlusToPublicKeys,
	validateSignature,
	verifyMinRemainingBalance,
	validateSenderIdAndPublicKey,
	verifyMultiSignatureTransaction,
};

export {
	Account,
	BaseTransaction,
	getSigningBytes,
	StateStore,
	TransferTransaction,
	DelegateTransaction,
	VoteTransaction,
	MultisignatureTransaction,
	UnlockTransaction,
	createResponse,
	ProofOfMisbehaviorTransaction,
	signTransaction,
	signMultiSignatureTransaction,
	Status,
	TransactionResponse,
	TransactionJSON,
	TransactionError,
	validateTransactionSchema,
	transactionInterface,
	convertToAssetError,
	convertToTransactionError,
	constants,
	exposedUtils as utils,
};
