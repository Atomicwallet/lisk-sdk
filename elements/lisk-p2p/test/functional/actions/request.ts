/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 *
 */
import { P2P, EVENT_REQUEST_RECEIVED } from '../../../src/index';
import {
	createNetwork,
	destroyNetwork,
	NETWORK_PEER_COUNT,
} from 'utils/network_setup';

describe('P2P.request', () => {
	let p2pNodeList: ReadonlyArray<P2P> = [];

	beforeEach(async () => {
		p2pNodeList = await createNetwork();

		for (let p2p of p2pNodeList) {
			// Collect port numbers to check which peer handled which request.
			p2p.on(EVENT_REQUEST_RECEIVED, request => {
				if (!request.wasResponseSent) {
					request.end({
						nodePort: p2p.nodeInfo.wsPort,
						requestProcedure: request.procedure,
						requestData: request.data,
						requestPeerId: request.peerId,
					});
				}
			});
		}
	});

	afterEach(async () => {
		await destroyNetwork(p2pNodeList);
	});

	it('should make request to the network; it should reach a single peer', async () => {
		const secondP2PNode = p2pNodeList[1];
		const response = await secondP2PNode.request({
			procedure: 'foo',
			data: 'bar',
		});
		expect(response).toMatchSnapshot();
	});

	// Check for even distribution of requests across the network. Account for an error margin.
	it('requests made to the network should be distributed randomly', async () => {
		const TOTAL_REQUESTS = 1000;
		const lastP2PNode = p2pNodeList[NETWORK_PEER_COUNT - 1];
		const { outboundCount } = lastP2PNode['_peerPool'].getPeersCountPerKind();
		const nodePortToResponsesMap: any = {};

		const expectedAverageRequestsPerNode = TOTAL_REQUESTS / outboundCount;
		const expectedRequestsLowerBound = expectedAverageRequestsPerNode * 0.5;
		const expectedRequestsUpperBound = expectedAverageRequestsPerNode * 1.5;

		for (let i = 0; i < TOTAL_REQUESTS; i++) {
			const response = await lastP2PNode.request({
				procedure: 'foo',
				data: i,
			});
			let resultData = response.data as any;
			if (!nodePortToResponsesMap[resultData.nodePort]) {
				nodePortToResponsesMap[resultData.nodePort] = [];
			}
			nodePortToResponsesMap[resultData.nodePort].push(resultData);
		}

		for (let requestsHandled of Object.values(nodePortToResponsesMap) as any) {
			expect(requestsHandled).toBeInstanceOf('array');
			expect(requestsHandled.length).toBeGreaterThan(
				expectedRequestsLowerBound,
			);
			expect(requestsHandled.length).toBeLessThan(expectedRequestsUpperBound);
		}
	}).timeout(5000);
});
