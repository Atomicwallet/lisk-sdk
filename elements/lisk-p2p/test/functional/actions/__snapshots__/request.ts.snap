// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`P2P.request should make request to the network; it should reach a single peer 1`] = `
Object {
  "data": Object {
    "nodePort": 5000,
    "requestData": "bar",
    "requestPeerId": "127.0.0.1:5001",
    "requestProcedure": "foo",
  },
  "peerId": "127.0.0.1:5001",
}
`;
